core/federated/federate.c

```diff
@@ -2835,53 +2531,301 @@ tag_t _lf_send_next_event_tag(environment_t* env, tag_t tag, bool wait_for_reply
-            federation_metadata.rti_host = rti_host;
+void lf_send_port_absent_to_federate(
+        environment_t* env,
+        interval_t additional_delay,
+        unsigned short port_ID,
+        unsigned short fed_ID) {
+    assert(env != GLOBAL_ENVIRONMENT);
+
+    // Construct the message
+    size_t message_length = 1 + sizeof(port_ID) + sizeof(fed_ID) + sizeof(instant_t) + sizeof(microstep_t);
+    unsigned char buffer[message_length];
+
+    // Apply the additional delay to the current tag and use that as the intended
+    // tag of the outgoing message. Note that if there is delay on the connection,
+    // then we cannot promise no message with tag = current_tag + delay because a
+    // subsequent reaction might produce such a message. But we can promise no
+    // message with a tag strictly less than current_tag + delay.
+    tag_t current_message_intended_tag = lf_delay_strict(env->current_tag, additional_delay);
+
+    LF_PRINT_LOG("Sending port "
+            "absent for tag " PRINTF_TAG " for port %d to federate %d.",
+            current_message_intended_tag.time - start_time,
+            current_message_intended_tag.microstep,
+            port_ID, fed_ID);
+
+    buffer[0] = MSG_TYPE_PORT_ABSENT;
+    encode_uint16(port_ID, &(buffer[1]));
+    encode_uint16(fed_ID, &(buffer[1+sizeof(port_ID)]));
+    encode_tag(&(buffer[1+sizeof(port_ID)+sizeof(fed_ID)]), current_message_intended_tag);
+
+#ifdef FEDERATED_CENTRALIZED
+    // Send the absent message through the RTI
+    int* socket = &_fed.socket_TCP_RTI;
+#else
+    // Send the absent message directly to the federate
+    int* socket = &_fed.sockets_for_outbound_p2p_connections[fed_ID];
+#endif
+
+    LF_MUTEX_LOCK(lf_outbound_socket_mutex);
+    int result = write_to_socket_close_on_error(socket, message_length, buffer);
+    LF_MUTEX_UNLOCK(lf_outbound_socket_mutex);
+
+    if (result != 0) {
+        // Write failed. Response depends on whether coordination is centralized.
+        if (socket == &_fed.socket_TCP_RTI) {
+            // Centralized coordination. This is a critical error.
+            lf_print_error_system_failure("Failed to send port absent message for port %hu to federate %hu.",
+                    port_ID, fed_ID);
         } else {
-            return INVALID_HOST;
+            // Decentralized coordination. This is not a critical error.
+            lf_print_warning("Failed to send port absent message for port %hu to federate %hu.",
+                    port_ID, fed_ID);
         }
-    }
-    if (rti_addr_info.has_port) {
-        if (validate_port(rti_addr_info.rti_port_str)) {
-            federation_metadata.rti_port = atoi(rti_addr_info.rti_port_str);
+    } else {
+        // Message sent correctly. Trace it.
+        if (socket == &_fed.socket_TCP_RTI) {
+            tracepoint_federate_to_rti(
+                    _fed.trace, send_PORT_ABS, _lf_my_fed_id, &current_message_intended_tag);
```

Comment:
FYI: **This isn't important yet**, but it may later become important that tracepoints are registered **after** the inputs of the traced event have been read but **before** the effects of the traced event have occurred. I have been thinking of this as the "input-tracepoint-effect assumption" and relying on it as a simplifying assumption when analyzing traces from federates running on the same node whose physical times come from a common central clock. Using this assumption it is possible to use the order in which two events happen to refute that a given event causally influences another event via a particular mechanism.

This is just a heads up in case it matters later or in case anyone runs into weird artifacts from a tracepoint being placed after the corresponding write to the network.

Reply:
This is a good point.  In commit ea398c78e3545c8649cb79abc12d96a3e614ef1d I have ensured that all tracepoints occur after socket reads and before socket writes.